microsoft/testsuites/core/cpu.py

```diff
@@ -0,0 +1,71 @@
+from lisa import testsuite
+from lisa.base_tools.cat import Cat
+from lisa.base_tools.uname import KernelVersion, Uname
+from lisa.node import Node
+from lisa.tools.lscpu import Lscpu
+from lisa.util import SkippedException
+
+NUMA_OFF = "numa=off"
+
+
+@testsuite.TestSuiteMetadata(
+    area="cpu",
+    category="functional",
+    description="""
+    This test suite is used to run CPU related tests.
+    """,
+)
+class CPU(testsuite.TestSuite):
+    @testsuite.TestCaseMetadata(
+        description="""
+        This test case will check that L3 cache is correctly mapped
+        to NUMA node.
+        Steps:
+        1. Check if NUMA is disabled in commandline. If disabled,
+        and kernel version is <= 2.6.37, test is skipped as kernel
+        has no support for NUMA.
+        2. Get the mappings using command :
+        `lscpu --extended=cpu,node,socket,cache`
+        3. Each line in the mapping corresponds to one CPU core. The L3
+        cache of each core must be mapped to the NUMA node that core
+        belongs to instead of the core itself.
+
+        Example :
+        Correct mapping:
+        CPU NODE SOCKET L1d L1i L2 L3
+        8   0    0      8   8   8  0
+        9   1    1      9   9   9  1
+
+        Incorrect mapping:
+        CPU NODE SOCKET L1d L1i L2 L3
+        8   0    0      8   8   8  8
+        9   1    1      9   9   9  9
+        """,
+        priority=0,
+    )
+    def l3_cache_check(self, node: Node) -> None:
+        cmdline = node.tools[Cat].run("/proc/cmdline").stdout
+        if NUMA_OFF in cmdline:
+            kernel_version: KernelVersion = (
+                node.tools[Uname].get_linux_information().kernel_version
+            )
+            self.__log.debug(
+                "Found numa=off in /proc/cmdline. Checking the kernel version."
+            )
+            if (
+                kernel_version.major <= 2
+                and kernel_version.minor <= 6
+                and kernel_version.patch <= 37
+            ):
+                raise SkippedException(
+                    "kernel : {kernel_version} has numa=off in boot "
+                    "parameter and its kernel version is earlier than 2.6.37. "
+                    "No support for NUMA setting. https://t.ly/x8k3"
+                )
+
+        cpu_info = node.tools[Lscpu].get_cpu_info()
+        for cpu in cpu_info:
+            assert cpu.node == cpu.l3, (
```

Comment:
use assert_that. That will show both values, if it's failed. but the original assert does not.

Reply:
I added a custom error message to show both values. But I agree that assert_that would be a cleaner option.

Reply:
BTW, the l3 looks like 13... how about level_3 or level_3_cache? full name is better.

Reply:
I will rename it to l3_cache and similar for other variables. That should be cleaner.